src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs

```diff
@@ -91,11 +90,20 @@ public override MatchProcessor Create(MatchProcessorReference matchProcessorRefe
 
         private MatchProcessor CreateMatchProcessorFromRouteConstraint(
             string parameterName,
+            bool optional,
```

Comment:
This is a bug. The MatchProcessorFactory doesn't wrap `IRouteConstraint` in optional.